File: android/app/src/main/kotlin/com/example/liveness_detection/logic/AntiSpoofGuard.kt
package com.example.liveness_detection.logic

import android.graphics.RectF
import android.util.Log
import com.example.liveness_detection.core.FaceLandmarks
import com.example.liveness_detection.core.ILandmarkExtractor
import kotlin.math.abs

/**
 * Result of anti-spoof check.
 */
sealed class AntiSpoofResult {
    data object Pass : AntiSpoofResult()
    data class Suspicious(val reason: String, val confidence: Float) : AntiSpoofResult()
    data class Spoof(val reason: String) : AntiSpoofResult()
}

/**
 * Passive liveness guard using geometric consistency checks.
 * 
 * Key insight: In a real 3D face, when the face moves closer/farther,
 * the bounding box size and inter-ocular distance change proportionally.
 * In a 2D screen (spoof), this ratio often becomes inconsistent due to
 * lens distortion and flat surface behavior.
 */
class AntiSpoofGuard(
    private val landmarkExtractor: ILandmarkExtractor
) {
    companion object {
        private const val TAG = "AntiSpoofGuard"
        
        // Minimum samples needed for analysis
        private const val MIN_SAMPLES = 10
        
        // Maximum allowed variance in box-to-IOD ratio
        private const val MAX_RATIO_VARIANCE = 0.15f
        
        // Minimum movement required for analysis (bbox change)
        private const val MIN_MOVEMENT_THRESHOLD = 0.05f
        
        // Suspicion threshold before marking as spoof
        private const val SUSPICION_THRESHOLD = 0.7f
    }
    
    // History of measurements for geometric consistency check
    private val measurements = mutableListOf<Measurement>()
    
    data class Measurement(
        val boundingBoxArea: Float,
        val interOcularDistance: Float,
        val timestamp: Long
    )
    
    /**
     * Add a new frame measurement for analysis.
     */
    fun addMeasurement(
        boundingBox: RectF,
        landmarks: FaceLandmarks
    ) {
        val boxArea = boundingBox.width() * boundingBox.height()
        val iod = landmarkExtractor.calculateInterOcularDistance(landmarks)
        
        if (iod > 0) {
            measurements.add(Measurement(
                boundingBoxArea = boxArea,
                interOcularDistance = iod,
                timestamp = System.currentTimeMillis()
            ))
            
            // Keep only recent measurements (last 30)
            if (measurements.size > 30) {
                measurements.removeAt(0)
            }
        }
    }
    
    /**
     * Analyze collected measurements for spoofing indicators.
     * 
     * The key check: In a real face, ratio of (BoundingBox Area) to 
     * (Inter-Ocular Distance²) should remain relatively constant
     * as the face moves toward/away from camera.
     */
    fun analyze(): AntiSpoofResult {
        if (measurements.size < MIN_SAMPLES) {
            Log.d(TAG, "Not enough samples for analysis: ${measurements.size}/$MIN_SAMPLES")
            return AntiSpoofResult.Pass // Not enough data yet
        }
        
        // Check if there's enough movement in the data
        val minBox = measurements.minOfOrNull { it.boundingBoxArea } ?: 0f
        val maxBox = measurements.maxOfOrNull { it.boundingBoxArea } ?: 0f
        val movementRange = if (maxBox > 0) (maxBox - minBox) / maxBox else 0f
        
        if (movementRange < MIN_MOVEMENT_THRESHOLD) {
            Log.d(TAG, "Not enough movement for analysis: $movementRange")
            return AntiSpoofResult.Pass // Face didn't move enough to analyze
        }
        
        // Calculate consistency ratio for each measurement
        // Ratio = BoundingBoxArea / (InterOcularDistance²)
        // This should be relatively constant for a real 3D face
        val ratios = measurements.map { m ->
            m.boundingBoxArea / (m.interOcularDistance * m.interOcularDistance)
        }
        
        val meanRatio = ratios.average().toFloat()
        val variance = ratios.map { abs(it - meanRatio) / meanRatio }.average().toFloat()
        
        Log.d(TAG, "Geometric analysis - Mean ratio: $meanRatio, Variance: $variance")
        
        return when {
            variance > MAX_RATIO_VARIANCE * 2 -> {
                Log.w(TAG, "SPOOF DETECTED - High variance: $variance")
                AntiSpoofResult.Spoof("Unnatural geometric behavior detected")
            }
            variance > MAX_RATIO_VARIANCE -> {
                Log.w(TAG, "Suspicious behavior - Variance: $variance")
                AntiSpoofResult.Suspicious(
                    "Unusual movement pattern",
                    variance / (MAX_RATIO_VARIANCE * 2)
                )
            }
            else -> {
                Log.d(TAG, "Geometric check passed")
                AntiSpoofResult.Pass
            }
        }
    }
    
    /**
     * Quick check for obvious 2D indicators.
     * Returns true if definitely a spoof, false if needs more analysis.
     */
    fun quickSpoofCheck(landmarks: FaceLandmarks): Boolean {
        // Check for unnaturally flat z-coordinates
        // In a real face, there's significant z-depth variation
        val zValues = landmarks.landmarks.map { it.z }
        val zRange = (zValues.maxOrNull() ?: 0f) - (zValues.minOrNull() ?: 0f)
        
        // If z-range is near zero, likely a flat image
        if (zRange < 0.01f) {
            Log.w(TAG, "Quick spoof check: Flat z-profile detected")
            return true
        }
        
        return false
    }
    
    /**
     * Reset all measurements.
     */
    fun reset() {
        measurements.clear()
        Log.d(TAG, "AntiSpoofGuard reset")
    }
    
    /**
     * Get current sample count.
     */
    fun getSampleCount(): Int = measurements.size
}

File: ios/Runner/Logic/AntiSpoofGuard.swift
import Foundation
import CoreGraphics

/// Result of anti-spoof check.
enum AntiSpoofResult {
    case pass
    case suspicious(reason: String, confidence: Float)
    case spoof(reason: String)
}

/// Passive liveness guard using geometric consistency checks.
/// Matches Android AntiSpoofGuard functionality.
///
/// Key insight: In a real 3D face, when the face moves closer/farther,
/// the bounding box size and inter-ocular distance change proportionally.
/// In a 2D screen (spoof), this ratio often becomes inconsistent due to
/// lens distortion and flat surface behavior.
class AntiSpoofGuard {
    
    // Minimum samples needed for analysis
    private let minSamples = 10
    
    // Maximum allowed variance in box-to-IOD ratio
    private let maxRatioVariance: Float = 0.15
    
    // Minimum movement required for analysis (bbox change)
    private let minMovementThreshold: Float = 0.05
    
    // Suspicion threshold before marking as spoof
    private let suspicionThreshold: Float = 0.7
    
    // History of measurements for geometric consistency check
    private var measurements: [Measurement] = []
    
    struct Measurement {
        let boundingBoxArea: Float
        let interOcularDistance: Float
        let timestamp: TimeInterval
    }
    
    private let landmarkExtractor: LandmarkExtractorProtocol
    
    init(landmarkExtractor: LandmarkExtractorProtocol) {
        self.landmarkExtractor = landmarkExtractor
    }
    
    /// Add a new frame measurement for analysis.
    func addMeasurement(boundingBox: CGRect, landmarks: FaceLandmarks) {
        let boxArea = Float(boundingBox.width * boundingBox.height)
        let iod = landmarkExtractor.calculateInterOcularDistance(landmarks: landmarks)
        
        if iod > 0 {
            measurements.append(Measurement(
                boundingBoxArea: boxArea,
                interOcularDistance: iod,
                timestamp: Date().timeIntervalSince1970
            ))
            
            // Keep only recent measurements (last 30)
            if measurements.count > 30 {
                measurements.removeFirst()
            }
        }
    }
    
    /// Analyze collected measurements for spoofing indicators.
    ///
    /// The key check: In a real face, ratio of (BoundingBox Area) to
    /// (Inter-Ocular Distance²) should remain relatively constant
    /// as the face moves toward/away from camera.
    func analyze() -> AntiSpoofResult {
        if measurements.count < minSamples {
            print("[AntiSpoofGuard] Not enough samples for analysis: \(measurements.count)/\(minSamples)")
            return .pass  // Not enough data yet
        }
        
        // Check if there's enough movement in the data
        let minBox = measurements.map { $0.boundingBoxArea }.min() ?? 0
        let maxBox = measurements.map { $0.boundingBoxArea }.max() ?? 0
        let movementRange = maxBox > 0 ? (maxBox - minBox) / maxBox : 0
        
        if movementRange < minMovementThreshold {
            print("[AntiSpoofGuard] Not enough movement for analysis: \(movementRange)")
            return .pass  // Face didn't move enough to analyze
        }
        
        // Calculate consistency ratio for each measurement
        // Ratio = BoundingBoxArea / (InterOcularDistance²)
        // This should be relatively constant for a real 3D face
        let ratios = measurements.map { m -> Float in
            m.boundingBoxArea / (m.interOcularDistance * m.interOcularDistance)
        }
        
        let meanRatio = ratios.reduce(0, +) / Float(ratios.count)
        let variance = ratios.map { abs($0 - meanRatio) / meanRatio }.reduce(0, +) / Float(ratios.count)
        
        print("[AntiSpoofGuard] Geometric analysis - Mean ratio: \(meanRatio), Variance: \(variance)")
        
        if variance > maxRatioVariance * 2 {
            print("[AntiSpoofGuard] SPOOF DETECTED - High variance: \(variance)")
            return .spoof(reason: "Unnatural geometric behavior detected")
        } else if variance > maxRatioVariance {
            print("[AntiSpoofGuard] Suspicious behavior - Variance: \(variance)")
            return .suspicious(
                reason: "Unusual movement pattern",
                confidence: variance / (maxRatioVariance * 2)
            )
        } else {
            print("[AntiSpoofGuard] Geometric check passed")
            return .pass
        }
    }
    
    /// Quick check for obvious 2D indicators.
    /// Returns true if definitely a spoof, false if needs more analysis.
    func quickSpoofCheck(landmarks: FaceLandmarks) -> Bool {
        // Check for unnaturally flat z-coordinates
        // In a real face, there's significant z-depth variation
        let zValues = landmarks.landmarks.map { $0.z }
        let zRange = (zValues.max() ?? 0) - (zValues.min() ?? 0)
        
        // If z-range is near zero, likely a flat image
        if zRange < 0.01 {
            print("[AntiSpoofGuard] Quick spoof check: Flat z-profile detected")
            return true
        }
        
        return false
    }
    
    /// Reset all measurements.
    func reset() {
        measurements.removeAll()
        print("[AntiSpoofGuard] AntiSpoofGuard reset")
    }
    
    /// Get current sample count.
    func getSampleCount() -> Int {
        return measurements.count
    }
}
